src/Data/List/Properties.agda

```diff
@@ -255,6 +255,27 @@ module _ (A : Set a) where
     ; ε-homo = refl
     }
 
+------------------------------------------------------------------------
+-- cartesianProductWith
+
+module _ (f : A → B → C) where
+
+  cartesianProductWith-zeroˡ : ∀ ys → cartesianProductWith f [] ys ≡ []
+  cartesianProductWith-zeroˡ _ = refl
+
+  cartesianProductWith-zeroʳ : ∀ xs → cartesianProductWith f xs [] ≡ []
+  cartesianProductWith-zeroʳ [] = refl
+  cartesianProductWith-zeroʳ (x ∷ xs) = cartesianProductWith-zeroʳ xs
+
+  cartesianProductWith-distribʳ-++ : ∀ xs ys zs → cartesianProductWith f (xs ++ ys) zs ≡ cartesianProductWith f xs zs ++ cartesianProductWith f ys zs
+  cartesianProductWith-distribʳ-++ [] ys zs = refl
+  cartesianProductWith-distribʳ-++ (x ∷ xs) ys zs = begin
+    cartesianProductWith f (x ∷ xs ++ ys) zs ≡⟨⟩
+    map (f x) zs ++ cartesianProductWith f (xs ++ ys) zs ≡⟨ cong (map (f x) zs ++_) (cartesianProductWith-distribʳ-++ xs ys zs) ⟩
+    map (f x) zs ++ cartesianProductWith f xs zs ++ cartesianProductWith f ys zs ≡˘⟨ ++-assoc (map (f x) zs) (cartesianProductWith f xs zs) (cartesianProductWith f ys zs) ⟩
+    (map (f x) zs ++ cartesianProductWith f xs zs) ++ cartesianProductWith f ys zs ≡⟨⟩
+    cartesianProductWith f (x ∷ xs) zs ++ cartesianProductWith f ys zs ∎
```

Comment:
Hmm this proof is very hard to read. The long name is definitely getting in the way. Maybe if we introduced a `private prod = cartesianProductWith` declaration at the top of this block it would improve readability? (while keeping the lemma names the same obviously!)

Also could you try and align the arguments as per the [style guide](https://github.com/agda/agda-stdlib/blob/master/notes/style-guide.md#function-arguments)?

Reply:
Hmm, if only this proof is hard to read wouldn't it be better if we introduce a `let` or `where` binding that is only local to this proof?

Also, wouldn't it be better to introduce the binding as `prod = cartesianProductWith f`?

Reply:
> Hmm, if only this proof is hard to read wouldn't it be better if we introduce a let or where binding that is only local to this proof?

The difficulty is that you can't make a `let` or `where` binding for which both the type and the body of the proof is in scope....

> Also, wouldn't it be better to introduce the binding as prod = cartesianProductWith f?

Yup, it would!

Reply:
You can use a `private` local definition in an anonymous module to have a declaration
span both the type & body of a proof.

Reply:
Yes, that was what I was suggesting.

Reply:
My bad !